### PR TITLE
basic tests using shared fixture suite, improve completeness & parity with JS

### PR DESCRIPTION
**Pull request by rvagg**
Sorry this is so large, I got deep into this once I started. The aim here was to bring this into closer alignment with current spec and to make it pass the test suite I'd been building at https://github.com/rvagg/js-ipld-schema — which I'd done using YAML & JSON to make it able to be portable when we had more than just a JS parser to test.

The majority of changes in here are done to get JSON marshalling working properly and matching spec. Some of that is awkward and involves replacing interfaces with concrete containers of subtypes (the representations do a lot of this), or the way `kind` is needed on everything, so it has to be supplied somewhere. parsing.go has most of the changes (note it'll be hidden by default by GitHub because the diff is large). schema.go will have some things that I'm sure will make you cringe but are there for JSONification.

Most of the testing involves JSONification, so it exercises that full path without caring about intermediate representation. The schema-schema test does JSON and back again to get matching `map[string]interface{}`s to compare.

I haven't done any tests yet for the to-schema path but I'd like to add that later to test that a schema parsed will go back the same way (it looks good atm, just eyeballing it). I also haven't touched anything in gen-go but it still compiles and runs at least.

If this doesn't match what you were planning on doing with this library then I'm find with doing something separate. Otherwise, perhaps we can work on getting this ready to go into github.com/ipld/ as the official Go parser? Either way, I've enjoyed this process and it's helped me think more Go-like and understand @warpfork's schema thinking a little more.

### CODE REVIEW

File: parser/parsing.go
Comment (whyrusleeping) [inline parser/parsing.go:511]: it seems like TypeStruct always has Kind "struct", and TypeList always has kind "list". maybe we can just encode that into the type somehow?

File: schema/schema.go
Comment (whyrusleeping) [inline schema/schema.go:10]: Ah, I see why you add that type there...

We could also implement the JSONMarshaler interface by adding a `MarshalJSON()` method to this type: https://golang.org/pkg/encoding/json/#Marshaler

example: https://golang.org/pkg/encoding/json/#example__customMarshalJSON

File: schema/schema.go
Comment (whyrusleeping) [inline schema/schema.go:213]: ooof, this is because of json parsing? Unfortunate, but i guess it works.

**Review by whyrusleeping — APPROVED** [review]
Oh wow those text fixtures. If those tests pass i'm happy with whatever, nothing stood out as obviously bad, just some things that go makes hard to do well. If you want to get @warpfork to look over this before merging that would be good, also fine to merge as is if you like

File: schema/schema.go
Comment (warpfork) [reply schema/schema.go:10]: Would rather try to reach escape velocity with our own codegen -- replacing this with output code that comes from  our own schema-schema will just insta-solve this, since it'll then have all its own JSON marshalling via the IPLD-standard mechanisms.

**Review by warpfork — APPROVED** [review]
Looked it over earlier, looks great.

Also had most of the same "ow" moments about the things that are in service to stdlib JSON convenience, heh.  We'll probably break free of those concerns again soon, I hope.  But for now it's a set of very practical, if frustrating, choices.

Share the excitement about all those test fixtures.  Those are really excellent.

**Pull request merged**
